### ISSUE TITLE
What should be in profiling output?

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#1676]
I was surprised to learn that the profile for `r([1,2,3,4,5]).filter{r(1).do{true}}` doesn't contain information on how long the filtering took, or a breakdown of the function body of the filter.  (See #1613.)  Is that what we want?

@jdoliner -- when you get a chance, could you explain what is and is not profiled and the motivation for it?  This is obviously way less important than the release.

**Comment by webmasterkai** [comment]
What about the time it takes to parse the query in the driver and in the db? I have seen several cases of a reported duration of 10ms in the profile output but have it actually take over 150ms from .run() to the return function. It seems changes in the way the query is written (but not its functionality) keep the reported duration within a few ms but change the roundtrip time (up to 700ms) dramatically.

**Comment by danielmewes** [comment]
@webmasterkai Not sure how easy it is to do, but I agree that query parsing at least on the server side would make sense to include.
Out of curiosity: What is the query for which the way you write it makes such a big difference? Does it have any large arrays? (in that case it might be a side-effect of https://github.com/rethinkdb/rethinkdb/issues/2919)

The profile currently has quite a few omissions and problems, and definitely needs some work to be really reliable.

**Comment by webmasterkai** [comment]
The more I use rethink the more I end up moving application logic into the query. Subqueries, http requests and the power or ReQL allows huge amounts of processing to be contained into a query. The issue I have is when the query gets so large that it's >20Kb of (no data, just code) ReQL. It's probably high time I backup and try to split things up a bit. Still, I'll see if I can put together a big mess of code to demonstrate slow query-parsing and report back with it.

**Comment by danielmewes** [comment]
@webmasterkai those are huge queries indeed. I can see that that would take considerable time for query parsing (especially with the inefficient way we currently generate backtraces for queries on the server). If you can put something together for us to test that would be great.
I suspect it will become more efficient once we solve https://github.com/rethinkdb/rethinkdb/issues/2900 . It's not a trivial change unfortunately, so that won't make it in before version 2.1.

**Comment by coffeemug** [comment]
I think the current profiler output is quite complex, over-engineered, and isn't very helpful. I'd consider ripping out most of the code, and just adding basic query statistics for now.

**Comment by coffeemug** [comment]
Actually, another option would be to take it out all together and just add the output to the `jobs` table.

**Comment by danielmewes** [comment]
I don't think there's a need for ripping it out and I don't think putting it into the `jobs` table would make for a good user experience (you'd have to query the `jobs` table exactly at the time your query is running, or you'd miss the profile).

I'm on board with simplifying the profiler output though, and only keeping things that actually work properly.

Arguably the profiler is actually under-engineered, making certain things difficult to track through it (which is why we aren't tracking them and why it's output isn't as relevant as it should be for some queries). At some point we'll have to put some thought into it and figure out if we can fix it or if we should instead reduce its scope to have a less powerful but better working profiler.

**Comment by VeXocide** [comment]
I agree the `jobs` table isn't the right place for this.

How about introducing an `r.explain()` which takes a query, runs it, returning the profiling output similar to SQL's `EXPLAIN`?

**Comment by coffeemug** [comment]
I think an optarg to `run` (like we do it now) is nicer than `r.explain`. It somehow feels... cleaner...

It's weird to add a term to do something that feels like a runtime option. Also, `explain` is quite different from profiling. In SQL, `EXPLAIN` looks at the query, the statistics, and then tells you how it will run the query. Profiling doesn't do that -- it actually runs the query, and tells you statistics about how it ran it. So I think we should keep them separate.

A few more thoughts:

- I now agree that putting this info into `jobs` is a bad idea. In addition to usability issues, the `jobs` table is really designed for ops people, while profiling is designed for developers, so it makes sense to keep them separate.
- I think the profiler is under-engineered in ways that matter and over-engineered in ways that don't (i.e. it gives lots of information that isn't helpful, and little information that *is* helpful).

**Comment by VeXocide** [comment]
You're right, an optarg to `run` makes more sense.

**Comment by mlucy** [comment]
I agree we should rip the profiler out.  It provides almost no useful information right now, and it's a constant source of headaches while writing ReQL code.  There are also some problems with the way it's designed:

* It builds up the trace in a way totally disconnected from the way we compute backtraces.  This makes it very hard to figure out what parts of your query correspond to what parts of the trace.
* It makes it very hard to parallelize things in the query language.  There's a single `trace` object, and depth is managed by calling functions on it to enter new chunks of code, so if you spawn two blocking coroutines to do work then you can't profile that work.  (There's also no way to combine traces really.)
* It doesn't have a good way to reconcile work done on the shards with work done eagerly on the node parsing the query, so we throw away a lot of information when we shard.

I think that it's extremely unlikely we will get from here to a good profiler by tweaking the one we have.  I think we should rip it out, keep the most useful statistics (query latency, number of shards hit, etc.), and possibly write a profiler in the future that works by attaching durations to parts of the backtrace.

**Comment by coffeemug** [comment]
:+1:

**Comment by danielmewes** [comment]
> I think we should rip it out, keep the most useful statistics (query latency, number of shards hit, etc.)

That sounds reasonable. We could just return an object with these statistics instead of the current profile.

**Comment by coffeemug** [comment]
> We could just return an object with these statistics instead of the current profile.

Yes! Also, when memory stats are in, we could add that to the object (which I would be *extremely* interested in seeing).

**Comment by GeoffreyPlitt** [comment]
It's been a few years, are memory stats being added to profile=true at all?

**Comment by GeoffreyPlitt** [comment]
Or #reads?
